### DISCUSSION TITLE
ADMatDiffusion vs MatDiffusion

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#17864] [Q&A General] [answered]
Hello, 

I am new to moose and a little confused with the differences between ADMatDiffusion and MatDiffusion. ADMatDiffusion uses the automatic differentiation system to calculate the Jacobian and MatDiffusion calculates the Jacobian directly inside the kernel. In both systems you can very the diffusivity as a function of args a, b, c, ... but what about u? My guess is that if you wanted the diffusitivity to very with u or a coupled variable such as temperature you would need to use ADMatDiffusion, is this correct?

**Comment by GiudGiud** [comment, marked as answer]
Hello

I think for MatDiffusion, the derivative with respect to u or the non-linear variables of the diffusivity are actually already included in the kernel definition. You can have a look at MatDiffusionBase.h for the Jacobian computation. 

So with MatDiffusion, as long as you define the derivatives of the diffusivities in your materials, you could still get the perfect Jacobian.
If you don't want to work on implementing the derivatives of your material properties, just use the AD version. 

Best,
Guillaume
